File: src/svc/systemcontroller/client_unit.go
package systemcontroller

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"

	"gitea.com/town-os/town-os/src/systemd"
)

// --- Systemd ---

// ListUnits returns a paginated list of systemd service units.
func (c *SystemdClient) ListUnits(ctx context.Context, params ListParams) (_ *PageResult[UnitListEntry], err error) {
	resp, err := c.getClient(ctx, "systemd/units"+params.QueryString())
	if err != nil {
		return nil, fmt.Errorf("%w: ListUnits: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "GET", "systemd/units")
	}

	var page PageResult[UnitListEntry]
	return &page, json.NewDecoder(resp.Body).Decode(&page)
}

// SetUnitStatus applies a status action (start, stop, restart) to a systemd unit.
func (c *SystemdClient) SetUnitStatus(ctx context.Context, name string, action systemd.StatusAction) error {
	pr, pw := io.Pipe()
	go pipeEncode(pw, SetStatusRequest{Name: name, Action: action})

	return c.postClient(ctx, "systemd/status", pr)
}

// ListUnitsTree returns package service units grouped into a dependency tree.
func (c *SystemdClient) ListUnitsTree(ctx context.Context, params ListParams) (_ *UnitTreeResponse, err error) {
	resp, err := c.getClient(ctx, "systemd/units-tree"+params.QueryString())
	if err != nil {
		return nil, fmt.Errorf("%w: ListUnitsTree: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "GET", "systemd/units-tree")
	}

	var out UnitTreeResponse
	return &out, json.NewDecoder(resp.Body).Decode(&out)
}

// SetUnitStatusTree cascades a status action across a package and all its
// installed dependencies.
func (c *SystemdClient) SetUnitStatusTree(ctx context.Context, repo, name, version string, action systemd.StatusAction) error {
	pr, pw := io.Pipe()
	go pipeEncode(pw, TreeStatusRequest{Repo: repo, Name: name, Version: version, Action: action})

	return c.postClient(ctx, "systemd/status/tree", pr)
}

// LogReplay streams historical journal entries for a unit via server-sent
// events. The returned channel is closed when the replay completes.
func (c *SystemdClient) LogReplay(ctx context.Context, name string) (<-chan systemd.JournalEntry, error) {
	return c.streamLogReplay(ctx, "systemd/logs?unit="+url.QueryEscape(name), "systemd/logs")
}

// LogTail returns a page of recent journal entries with cursor-based pagination.
func (c *SystemdClient) LogTail(ctx context.Context, p systemd.LogTailParams) (_ systemd.LogTailResult, err error) {
	q := fmt.Sprintf("systemd/logs/tail?unit=%s&lines=%d", url.QueryEscape(p.Unit), p.Lines)
	return c.fetchLogTail(ctx, appendLogTailQuery(q, p), "systemd/logs/tail")
}

// LogReplayTree streams the merged journal history for every systemd unit
// in a package's dependency tree.
func (c *SystemdClient) LogReplayTree(ctx context.Context, repo, name, version string) (_ <-chan systemd.JournalEntry, err error) {
	q := fmt.Sprintf("systemd/logs/tree?repo=%s&name=%s&version=%s",
		url.QueryEscape(repo), url.QueryEscape(name), url.QueryEscape(version))
	return c.streamLogReplay(ctx, q, "systemd/logs/tree")
}

// LogTailTree returns a page of journal entries drawn from every unit in a
// package's dependency tree.
func (c *SystemdClient) LogTailTree(ctx context.Context, repo, name, version string, p systemd.LogTailParams) (systemd.LogTailResult, error) {
	q := fmt.Sprintf("systemd/logs/tree/tail?repo=%s&name=%s&version=%s&lines=%d",
		url.QueryEscape(repo), url.QueryEscape(name), url.QueryEscape(version), p.Lines)
	return c.fetchLogTail(ctx, appendLogTailQuery(q, p), "systemd/logs/tree/tail")
}

// appendLogTailQuery attaches the shared filter and pagination params onto
// the caller's base query string. The base string must already contain the
// unit selector (either `unit=` for the single-unit endpoint or
// `repo=&name=&version=` for the tree endpoint) plus `lines=`.
func appendLogTailQuery(q string, p systemd.LogTailParams) string {
	if p.BeforeCursor != "" {
		q = fmt.Sprintf("%s&before=%s", q, url.QueryEscape(p.BeforeCursor))
	}
	if p.AfterCursor != "" {
		q = fmt.Sprintf("%s&after=%s", q, url.QueryEscape(p.AfterCursor))
	}
	if p.Grep != "" {
		q = fmt.Sprintf("%s&grep=%s", q, url.QueryEscape(p.Grep))
	}
	if !p.Since.IsZero() {
		q = fmt.Sprintf("%s&since=%d", q, p.Since.Unix())
	}
	if !p.Until.IsZero() {
		q = fmt.Sprintf("%s&until=%d", q, p.Until.Unix())
	}
	if p.Priority > 0 {
		q = fmt.Sprintf("%s&priority=%d", q, p.Priority)
	}
	return q
}

// fetchLogTail is the shared transport for the tail endpoints: it issues
// the GET, decodes the JSON body, and surfaces problem-detail errors on
// non-200 responses.
func (c *SystemdClient) fetchLogTail(ctx context.Context, q, pathForErr string) (_ systemd.LogTailResult, err error) {
	resp, err := c.getClient(ctx, q)
	if err != nil {
		return systemd.LogTailResult{}, fmt.Errorf("%w: LogTail: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return systemd.LogTailResult{}, readProblemDetail(resp, "GET", pathForErr)
	}

	var result systemd.LogTailResult
	return result, json.NewDecoder(resp.Body).Decode(&result)
}

// streamLogReplay is the shared SSE transport for the single-unit and
// tree replay endpoints.
func (c *SystemdClient) streamLogReplay(ctx context.Context, q, pathForErr string) (_ <-chan systemd.JournalEntry, err error) {
	resp, err := c.getClient(ctx, q) //nolint:bodyclose // closed in goroutine below
	if err != nil {
		if resp != nil {
			err = errors.Join(err, resp.Body.Close())
		}
		return nil, fmt.Errorf("%w: LogReplay: %w", ErrHTTPRequest, err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetailAndClose(resp, "GET", pathForErr)
	}

	ch := make(chan systemd.JournalEntry)
	go func() {
		defer close(ch)
		defer func() {
			err = errors.Join(err, resp.Body.Close())
		}()

		scanner := bufio.NewScanner(resp.Body)
		for scanner.Scan() {
			line := scanner.Text()
			if !strings.HasPrefix(line, "data: ") {
				continue
			}

			var entry systemd.JournalEntry
			if err := json.NewDecoder(strings.NewReader(strings.TrimPrefix(line, "data: "))).Decode(&entry); err != nil {
				return
			}
			select {
			case ch <- entry:
			case <-ctx.Done():
				return
			}
		}
	}()

	return ch, nil
}

File: ui/src/api/client-unit.js
/** @import { UnitStatus, JournalEntry, StatusAction } from './types.js' */
import { ApiError, SystemControllerClient } from './core.js'

/**
 * @param {string} [sortBy]
 * @param {string} [sortOrder]
 * @param {number} [limit]
 * @param {number} [offset]
 * @param {string} [search]
 * @returns {Promise<{entries: UnitStatus[], has_more: boolean, total_pages: number}>}
 */
SystemControllerClient.prototype.listUnits = async function (sortBy, sortOrder, limit, offset, search) {
  const params = new URLSearchParams()
  if (sortBy) params.set('sort_by', sortBy)
  if (sortOrder) params.set('sort_order', sortOrder)
  if (limit) params.set('limit', String(limit))
  if (offset) params.set('offset', String(offset))
  if (search) params.set('search', search)
  const qs = params.toString()
  return this.getJSON(`/systemd/units${qs ? `?${qs}` : ''}`)
}

/**
 * @param {string} unit
 * @returns {AsyncGenerator<JournalEntry>}
 */
SystemControllerClient.prototype.logReplay = async function* (unit) {
  /** @type {HeadersInit} */
  const headers = {}
  if (this.token) {
    headers['Authorization'] = `Bearer ${this.token}`
  }
  const resp = await fetch(
    `${this.baseURL}/systemd/logs?unit=${encodeURIComponent(unit)}`,
    { headers },
  )
  if (resp.status !== 200) {
    const text = await resp.text().catch(() => '')
    throw new ApiError('GET', '/systemd/logs', resp.status, text)
  }

  const reader = resp.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  for (;;) {
    const { done, value } = await reader.read()
    if (done) break

    buffer += decoder.decode(value, { stream: true })
    const lines = buffer.split('\n')
    buffer = lines.pop() ?? ''

    for (const line of lines) {
      if (line.startsWith('data: ')) {
        yield JSON.parse(line.slice(6))
      }
    }
  }

  if (buffer.startsWith('data: ')) {
    yield JSON.parse(buffer.slice(6))
  }
}

/**
 * Fetch a page of journal log entries for a unit with cursor-based pagination.
 * @param {string} unit - Systemd unit name.
 * @param {number} [lines=100] - Maximum number of entries to return.
 * @param {string} [beforeCursor] - Return entries before this opaque cursor (for backward paging).
 * @param {string} [afterCursor] - Return entries after this opaque cursor (for forward paging).
 * @param {string} [grep] - Filter to entries whose message contains this substring.
 * @param {number} [since] - Only entries at or after this Unix timestamp (seconds).
 * @param {number} [until] - Only entries at or before this Unix timestamp (seconds).
 * @param {number} [priority] - Maximum syslog priority level to include; 0 disables filtering,
 *   values 1–7 include entries with priority <= the value (e.g. 3 returns emergency, alert,
 *   critical, and error).
 * @returns {Promise<{entries: JournalEntry[], cursor: string, end_cursor: string}>}
 */
SystemControllerClient.prototype.logTail = async function (unit, lines = 100, beforeCursor, afterCursor, grep, since, until, priority) {
  const params = new URLSearchParams({ unit, lines: String(lines) })
  if (beforeCursor) params.set('before', beforeCursor)
  if (afterCursor) params.set('after', afterCursor)
  if (grep) params.set('grep', grep)
  if (since) params.set('since', String(since))
  if (until) params.set('until', String(until))
  if (priority) params.set('priority', String(priority))
  return this.getJSON(`/systemd/logs/tail?${params.toString()}`)
}

/**
 * Apply a status action to a systemd unit.
 * @param {string} name - Systemd unit name.
 * @param {StatusAction} action - Action to apply: "start", "stop", or "restart".
 * @returns {Promise<void>}
 */
SystemControllerClient.prototype.setUnitStatus = async function (name, action) {
  await this.post('/systemd/status', { name, action })
}

/**
 * Fetch package service units grouped into a dependency tree.
 * @param {string} [sortBy]
 * @param {string} [sortOrder]
 * @param {number} [limit]
 * @param {number} [offset]
 * @param {string} [search]
 * @returns {Promise<{entries: any[], has_more: boolean, total_pages: number, total_count: number}>}
 */
SystemControllerClient.prototype.listUnitsTree = async function (sortBy, sortOrder, limit, offset, search) {
  const params = new URLSearchParams()
  if (sortBy) params.set('sort_by', sortBy)
  if (sortOrder) params.set('sort_order', sortOrder)
  if (limit) params.set('limit', String(limit))
  if (offset) params.set('offset', String(offset))
  if (search) params.set('search', search)
  const qs = params.toString()
  return this.getJSON(`/systemd/units-tree${qs ? `?${qs}` : ''}`)
}

/**
 * Cascade a status action across a package and all its installed dependencies.
 * @param {string} repo
 * @param {string} name - Raw effective package name (may contain "--dep--").
 * @param {string} version
 * @param {StatusAction} action - "start", "stop", or "restart".
 * @returns {Promise<void>}
 */
SystemControllerClient.prototype.setUnitStatusTree = async function (repo, name, version, action) {
  await this.post('/systemd/status/tree', { repo, name, version, action })
}

/**
 * Stream historical journal entries for every systemd unit in a package's
 * dependency tree as a single SSE stream. Entries from the parent and
 * every dependency arrive interleaved in chronological order.
 * @param {string} repo
 * @param {string} name - Raw effective package name (may contain "--dep--").
 * @param {string} version
 * @returns {AsyncGenerator<JournalEntry>}
 */
SystemControllerClient.prototype.logReplayTree = async function* (repo, name, version) {
  /** @type {HeadersInit} */
  const headers = {}
  if (this.token) {
    headers['Authorization'] = `Bearer ${this.token}`
  }
  const params = new URLSearchParams({ repo, name, version })
  const resp = await fetch(`${this.baseURL}/systemd/logs/tree?${params.toString()}`, { headers })
  if (resp.status !== 200) {
    const text = await resp.text().catch(() => '')
    throw new ApiError('GET', '/systemd/logs/tree', resp.status, text)
  }

  const reader = resp.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  for (;;) {
    const { done, value } = await reader.read()
    if (done) break

    buffer += decoder.decode(value, { stream: true })
    const lines = buffer.split('\n')
    buffer = lines.pop() ?? ''

    for (const line of lines) {
      if (line.startsWith('data: ')) {
        yield JSON.parse(line.slice(6))
      }
    }
  }

  if (buffer.startsWith('data: ')) {
    yield JSON.parse(buffer.slice(6))
  }
}

/**
 * Fetch a page of journal entries covering every systemd unit in a
 * package's dependency tree. Filters and cursors behave identically to
 * logTail; the unit set is derived server-side from the install records.
 * @param {string} repo
 * @param {string} name - Raw effective package name (may contain "--dep--").
 * @param {string} version
 * @param {number} [lines=100]
 * @param {string} [beforeCursor]
 * @param {string} [afterCursor]
 * @param {string} [grep]
 * @param {number} [since]
 * @param {number} [until]
 * @param {number} [priority]
 * @returns {Promise<{entries: JournalEntry[], cursor: string, end_cursor: string}>}
 */
SystemControllerClient.prototype.logTailTree = async function (repo, name, version, lines = 100, beforeCursor, afterCursor, grep, since, until, priority) {
  const params = new URLSearchParams({ repo, name, version, lines: String(lines) })
  if (beforeCursor) params.set('before', beforeCursor)
  if (afterCursor) params.set('after', afterCursor)
  if (grep) params.set('grep', grep)
  if (since) params.set('since', String(since))
  if (until) params.set('until', String(until))
  if (priority) params.set('priority', String(priority))
  return this.getJSON(`/systemd/logs/tree/tail?${params.toString()}`)
}
